lib/src/strings.dart

```diff
@@ -175,24 +232,19 @@ String yamlEncodeBlockScalar(
 
     final val = value.value;
     if (val is String) {
-      if (_hasUnprintableCharacters(val)) {
-        return _yamlEncodeDoubleQuoted(val);
-      }
-
-      if (value.style == ScalarStyle.SINGLE_QUOTED) {
-        return _tryYamlEncodeSingleQuoted(val);
-      }
-
-      // Strings with only white spaces will cause a misparsing
-      if (val.trim().length == val.length && val.isNotEmpty) {
-        if (value.style == ScalarStyle.FOLDED) {
-          return _tryYamlEncodeFolded(val, indentation, lineEnding);
-        }
-
-        if (value.style == ScalarStyle.LITERAL) {
-          return _tryYamlEncodeLiteral(val, indentation, lineEnding);
+      String? encoded;
+
+      if (!_hasUnprintableCharacters(val)) {
+        if (value.style == ScalarStyle.SINGLE_QUOTED) {
+          encoded = _tryYamlEncodeSingleQuoted(val);
+        } else if (value.style == ScalarStyle.FOLDED) {
+          encoded = _tryYamlEncodeFolded(val, indentation, lineEnding);
+        } else if (value.style == ScalarStyle.LITERAL) {
+          encoded = _tryYamlEncodeLiteral(val, indentation, lineEnding);
         }
       }
+
+      if (encoded != null) return encoded;
     }
 
     return _tryYamlEncodePlain(value.value);
```

Comment:
How about something like this?

We make all the `_try...` methods return `null` if they can't encode correctly.
And they we try the best option first, and just try a few relevant fallbacks.

It's probably rare that literal works when folded doesn't, and vice versa, but it can happen, right? And trying is probably rather cheap.

I realize that we do end up checking `_hasUnprintableCharacters` inside each method. But I'm guess it's fairly cheap, and we'll only run it more than once in the cases where there actually is something unprintable. So it won't really affect performance much (after all, it's rare that there are unprintable characters).

```suggestion
      switch (value.style) {
        // Prefer 'plain', fallback to "double quoted"
        case ScalarStyle.PLAIN:
          return  _tryYamlEncodePlain(value.value) ?? _yamlEncodeDoubleQuoted(val);
        
        // Prefer 'single quoted', fallback to "double quoted"
        case ScalarStyle.SINGLE_QUOTED:
          return _tryYamlEncodeSingleQuoted(val) ?? _yamlEncodeDoubleQuoted(val);

        // Prefer folded string, try literal as fallback
        // otherwise fallback to "double quoted"
        case ScalarStyle.FOLDED:
          return _tryYamlEncodeFolded(val, indentation, lineEnding)
            ?? _tryYamlEncodeLiteral(val, indentation, lineEnding)
            ?? _yamlEncodeDoubleQuoted(val);

        // Prefer literal string, try folded as fallback
        // otherwise fallback to "double quoted"
        case ScalarStyle.LITERAL:
          return _tryYamlEncodeLiteral(val, indentation, lineEnding)
            ?? _tryYamlEncodeFolded(val, indentation, lineEnding)
            ?? _yamlEncodeDoubleQuoted(val);

        // Prefer plain, try 'single quote', fallback to "double quoted"
        default:
          return _tryYamlEncodePlain(value.value)
            ?? _tryYamlEncodeSingleQuoted(val)
            ?? _yamlEncodeDoubleQuoted(val);
      }
    }
```

Reply:
`_tryYamlEncodePlain` internally falls back to `_yamlEncodeDoubleQuoted`. Should I make it return `null` if it fails instead?

Reply:
Yeah, let's, then it'll be consistent.

If we do a switch statement like above, we get to deal with one thing in each method. And the heuristics of what we fall back to is all located in the switch statement.